danmaku/ui/commonMain/DanmakuHostState.kt

```diff
@@ -0,0 +1,393 @@
+                speedPxPerSecond = newFloatingTrackSpeed,
+                safeSeparation = newFloatingTrackSafeSeparation,
+                baseTextLength = floatingBaseTextLengthForSpeed,
+                speedMultiplier = floatingSpeedMultiplierState,
+                onRemoveDanmaku = { removed -> mutablePresentDanmaku.remove(removed) }
+            )
+        }
+        floatingTrack.forEach {
+            it.speedPxPerSecond = newFloatingTrackSpeed
+            it.safeSeparation = newFloatingTrackSafeSeparation
+            it.baseTextLength = floatingBaseTextLengthForSpeed
+        }
+        topTrack.setTrackCountImpl(if (config.enableTop) count else 0) { index ->
+            FixedDanmakuTrack(
+                trackIndex = index,
+                frameTimeNanos = elapsedFrameTimeNanoState,
+                trackHeight = trackHeightState,
+                screenWidth = hostWidthState,
+                screenHeight = hostHeightState,
+                fromBottom = false,
+                onRemoveDanmaku = { removed -> mutablePresentDanmaku.remove(removed) }
+            )
+        }
+        bottomTrack.setTrackCountImpl(if (config.enableBottom) count else 0) { index ->
+            FixedDanmakuTrack(
+                trackIndex = index,
+                frameTimeNanos = elapsedFrameTimeNanoState,
+                trackHeight = trackHeightState,
+                screenWidth = hostWidthState,
+                screenHeight = hostHeightState,
+                fromBottom = true,
+                onRemoveDanmaku = { removed -> mutablePresentDanmaku.remove(removed) }
+            )
+        }
+        invalidate()
+    }
+
+    private fun <T : DanmakuTrack> MutableList<T>.setTrackCountImpl(
+        count: Int,
+        newInstance: (index: Int) -> T,
+    ) {
+        when {
+            size == count -> return
+            // 清除 track 的同时要把 track 里的 danmaku 也要清除
+            count < size -> repeat(size - count) { removeLast().clearAll() }
+            else -> addAll(List(count - size) { newInstance(size + it) })
+        }
+    }
+
+    @UiThread
+    internal suspend fun interpolateFrameLoop() {
+        coroutineScope {
+            launch {
+                var currentFrameTimeNanos = withFrameNanos {
+                    // 使用了这一帧来获取时间, 需要补偿平均帧时间
+                    elapsedFrameTimeNanos += avgFrameTimeNanos.avg()
+                    it
+                }
+
+                while (true) {
+                    withFrameNanos { nanos ->
+                        val delta = nanos - currentFrameTimeNanos
+                        elapsedFrameTimeNanos += delta
+                        avgFrameTimeNanos += delta
+                        currentFrameTimeNanos = nanos
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * 逻辑帧 tick, 主要用于移除超出屏幕外或超过时间的弹幕
+     */
+    @UiThread
+    internal suspend fun tickLoop() {
+        coroutineScope { 
+            launch {
+                while (isActive) {
```

Comment:
```suggestion
                while (true) {
```

`delay` always check for cancellation